### ISSUE TITLE
Add native `dialog` element support

### ISSUE THREAD

**Issue by tay1orjones** [carbon-design-system/carbon#13807]
## Context

An initial exploration and delivery of a `Dialog` component was done a while ago: 
- https://github.com/carbon-design-system/carbon/issues/8826
- https://github.com/carbon-design-system/carbon/pull/9496

This experimental work was never exported or shown in storybook but still lives in the codebase. Most of the spec of that work stream was focused on managing focus and handling interactions such as <kbd>escape</kbd> closing the dialog.

## The current state

Since then, there has been some progress on the native dialog element spec and much of this focus and interaction behavior is now provided natively by browsers. 

- https://www.scottohara.me/blog/2023/01/26/use-the-dialog-element.html
- https://www.youtube.com/watch?v=ywtkJkxJsdg
- https://developer.mozilla.org/en-US/docs/Web/HTML/Element/dialog
- https://www.makethingsaccessible.com/guides/modal-vs-non-modal-dialogs

## Proposal

This issue is to propose considering using the native dialog element in place of the existing implementation that uses the `FocusScope` primitive.

Depending on what we consider to be "must have" functionality of a `Dialog`, it might be that the bulk of the requirements might be met by the `dialog` element. It might be a simple wrapper around that element is all we need.

### Pieces solved in https://github.com/carbon-design-system/carbon/pull/15926
- [x] Evaluate the interaction provided via the native element
- [x] Compare the native implementation to the one we have in the codebase today
- [x] Determine what we may need to provide to consumers in addition to what is provided by default by the browser implementation
- [x] Potentially refactor the implementation to not manage focus via the FocusScope helpers (potentially remove these helpers if they're not being used)

### Outstanding items that still need to be made into sub-issues
- [ ] Add Dialog and associated refactors/usages/etc. to `@carbon/web-components`
- [ ] Move to stable , work through stable checklist (code, website docs, etc.)
- [ ] ...

**Comment by mbgower** [comment]
Some considerations for the assessment:
- granularity for placing focus when dialog appears (on dialog element, on controls inside dialog)
- effect on scrolling inside dialog, depending on focus placement
- keyboard retention in modal dialog
- naming/title for the dialog

In addition, for the non-modal version of the dialog, the following should be assessed, to discover potential issues meeting WCAG 2.2's [Focus Not Obscured (Minimum)](https://w3c.github.io/wcag/understanding/focus-not-obscured-minimum.html) requirement as well as other existing requirements and conventions on focus handling, prior point of regard, focus indicators, etc. :
- dismissal mechanisms
- persistence on loss of focus
- positioning in DOM

**Comment by tay1orjones** [comment]
As part of https://github.com/carbon-design-system/carbon/pull/14156 @guidari and I tested the native dialog functionality and couldn't get the focus wrap behavior to work at all. I'm not sure if it's behind a flag for now or perhaps chrome/firefox hasn't implemented it yet fully?

Either way I think we're blocked until we can confirm the new native dialog behavior functions as expected in all supported browsers.

**Comment by tay1orjones** [comment]
We'll also need to evaluate and document how other [top layer](https://developer.chrome.com/blog/what-is-the-top-layer) elements interact with a `dialog`. Initial assumptions/questions I have:

* Portals will not work as expected. Elements rendered via portals will always render behind the dialog.
* What is the behavior of our Popover (Tooltip, Toggletip, etc) - do these render above or below the native dialog when placed within it?
* Does a Popover with autoAlign work as intended, is it the same as non-autoAlign popovers?
* If an element within the dialog uses the native Popover API, which element sits on top?
* If an element using the native Popover API renders a dialog, which element sits on top?

**Comment by lee-chase** [comment]
@tay1orjones it would appear that completely trapping focus is not desired. Tabbing to browser/environment controls is desirable, page content controls are blocked. 

 https://github.com/whatwg/html/issues/8339#issuecomment-1822591131

**Comment by lee-chase** [comment]
Use of the the native dialog component

 https://codepen.io/lee-chase/pen/QWPbRLp/9c8c3f23b42a395dc519839e12f5d647

**Comment by lee-chase** [comment]
Unfortunately the transition behavior to `allow-discrete` is not well supported and would need script

https://caniuse.com/?search=transition-behavior

**Comment by lee-chase** [comment]
Experimented and created this based on the native dialog.
https://github.com/carbon-design-system/carbon/pull/15926

**Comment by tay1orjones** [comment]
I've updated this issue with continued direction now that the initial exploration has been merged in. Logical next step would be to refactor modal behind a feature flag to use `Dialog`

> Unfortunately the transition behavior to `allow-discrete` is not well supported and would need script

In https://github.com/carbon-design-system/carbon/pull/15926 we landed on not trying to polyfill this. Since it doesn't break anything other than the animation, we can consider it a case of progressive enhancement. It works in chrome/edge now and should work in safari this fall since it's already in the TP. Firefox users just won't see the animation until it's implemented there.

**Comment by sstrubberg** [comment]
Testing
Rollout strategy
Flags
Modal, Composed Modal
Tearsheets, Side Panels(?), NotificationsPanel, Interstitial, TagSet.
Stacking in Tearsheets makes more sense in Dialogs than in Modals.
Check in with @lee-chase

**Comment by ljcarot** [comment]
@sstrubberg @tay1orjones if this is a prereq to TearSheet, how much more work is needed here?

**Comment by FrivalszkyP** [comment]
Hi folks! We're working on IBM Cloud accessibility issues at the moment and we found this github issue. We also found out that the WCAG rule 2.4.3 Focus Order is _technically_ not met by the browser native dialog, even though the working group members agree that it satisfies the desired behavior. A Scott O'Hara says in the https://github.com/whatwg/html/issues/8339#issuecomment-2328627897 issue, "the doc is still accurate for custom dialogs. Might help if you file an issue against wcag for the update to be considered."

I volunteer to work on putting together an issue for WCAG, or, even bolder, a PR to include references to the browser native dialog implementation. 

The reason I'd like to work on this is that our accessibility verification testing (understandably) adheres to the current WCAG requirements, which do state that the keyboard should be trapped within the modal dialog. If there was wording to address the native dialog behavior, that could also help the Carbon team move forward with making the native dialog-based Modal implementation GA. What do you folks think, are you interested in working with WCAG on this?

**Comment by sstrubberg** [comment]
@FrivalszkyP thanks for surfacing this. Native dialog support is something we are pushing hard for in Carbon. Having said that, making sure it meets WCAG spec is crucial for it's success. I'll get with the team and we'll see about a path forward.

**Comment by tay1orjones** [comment]
> the WCAG rule 2.4.3 Focus Order is _technically_ not met by the browser native dialog

Yeah, the technicality is browsers include the browser URL bar in the focus order trap behavior for native `<dialog>`s. When focus moves forward from the last element in the `<dialog>`, it goes to the URL bar. When focus moves forward again, it goes back into the `<dialog>`.

This is the correct and agreed-upon behavior, but WCAG was written much earlier for _custom_ dialogs (like our current Modal) before `<dialog>` and `inert` existed. 

@FrivalszkyP I found https://github.com/w3c/wcag/issues/4185 already exists for this. I just put a comment on there. If you want to shoot for a PR there, @mbgower and/or @tombrunet might have some thoughts as they're members of the WG.

**Comment by tay1orjones** [comment]
FWIW there might be more standards-based issues that come along preventing this from moving to stable. One example is https://github.com/whatwg/html/issues/9376

None of these would be hard blockers to our work or use of `<dialog>` in the meantime though. It's still abundantly clear native `<dialog>`s are the path forward.

**Comment by tombrunet** [comment]
@FrivalszkyP @tay1orjones There seems to be a huge misconception here. So, I'll be clear. There is nothing within WCAG that says that modal dialogs have to trap focus. The misconception is derived from examples of things that look like focus traps that aren't actually focus traps, and WCAG trying to explain why that isn't a keyboard trap.

So, all of the quoted items in those other issues are where WCAG is setting up a scenario that might be a violation, but then an explanation of why it isn't. In the case of a custom modal that traps focus within the webpage, it would be a violation of 2.1.2 IF it could not be dismissed via standard key navigation (typically "Escape" or a Cancel button). 

By allowing the user to tab directly to the browser, all of those examples are irrelevant because there is nothing that resembles a keyboard trap to begin with.

**Comment by tombrunet** [comment]
As an aside, 2.1.2 was largely created to deal with browser plug-ins that basically no longer exist. In those scenarios, you literally could not get back to the browser without touching your mouse. It still somewhat dissuades someone from creating a non-dismissable dialog, but the concept of keyboard trap as it existed when 2.1.2 was written no longer happens on the web as far as I've seen.
